docs/GENERATORS.md

```diff
@@ -0,0 +1,103 @@
+
+As an example, if you wanted to change the default behavior so that when the `Input` value of a test is a negative number, an exception should be thrown, the code would look like this.
+
+```csharp
+protected override void UpdateCanonicalData(CanonicalData canonicalData)
+{
+  foreach (var canonicalDataCase in canonicalData.Cases)
+  {
+    var caseInputLessThanZero = (long)canonicalDataCase.Input["number"] <= 0;
+    canonicalDataCase.ExceptionThrown = caseInputLessTanZero ? typeof(ArgumentException) : null;
+  }
+}
+```
+
+### HashSet\<string\> AddAdditionalNamespaces()
+Allows more namespaces to be added to the test suite. 
+
+The tests use `Xunit` so all tests will automatically include the `Xunit` namespace. However, more advanced tests will require additional namespaces.
+
+```csharp
+protected override HashSet<string> AddAdditionalNamespaces()
+{
+  return new HashSet<string>()
+  {
+    typeof(Dictionary<char, int>).Namespace
+  };
+}
+```
+
+This snippet would add the namespace that the `Dictionary<char,int>` collection lives in (`System.Collections.Generic`).
+
+### string RenderTestMethodBody[Arrange/Act/Assert]
+Override the default behavior when rendering a test methods arrange, act, and/or assert sections.
+
+More advanced tests may need to leverage a `template`. A template allows you to add additional code to a test and assert more complex statements.
+
+An example of this is the [RunLengthEncoding](https://github.com/exercism/csharp/blob/master/generators/Exercises/RunLengthEncoding.cs) test.
+
+Here the **Assert** is being overridden, as the Assert needs to call additional functions, but only if the property is consistency. Otherwise, render the assert as usual.
+
+## Updating Existing Files
+It is possible that an existing exercise does not match the canonical data. It is OK to update the exercise stub and/or the exercise example to follow the canonical data! An example might be that an exercise is named SumOfMultiples, but the SumOfMultiples.cs and Example.cs files both use `Multiples` as the name of the class.
+
+Also, if you find an issue with one of the existing generators or test suites simply open up the generator that you would like to update, make your changes, and then run the generators.
+
+## Running The Generators
+This repository is coded against .NET Core. To run all of the generators all you need to do is run the following command in the generators directory:
+
+`dotnet run generators`
+
+This command will take all of the generators that are in the `Exercises` folder, and generate all of the test cases for that exercise. We use reflection to get all of the exercises, so if you are adding a new test, the test will be automatically included when running the generator.
+
+Once the generator has been ran, you can view the output of your generation by navigating to the test file for that exercise. As an example, the test suite for the Bob exercise can be found at:
+
+`csharp/exercises/bob/BobTest.cs`
```

Comment:
I think you can omit the `csharp/` part, as people are probably already in the project's directory.